### PR TITLE
feat: create dummy grid element capability in datarecord

### PR DESCRIPTION
**Pull request by kbarnhart**
As part of the discussion regarding #909, and other needs of @pfeiffea and I with regard to the network sediment transporter, I'm replacing the now closed #948 with two PRs. This is the second of them (after #989). It branches from that dev branch and I expect that it will be merged _after_ #989. 

It adds the capability to store items on "dummy" grid elements. 

Right now its tests don't pass. I'm working on fixing them, and when I'm done I'll tag people to review.

### CODE REVIEW

File: landlab/data_record/data_record.py [hunk 1]
```diff
@@ -14,40 +15,70 @@ class DataRecord(object):
...
 
+    If an item or set of items is defined, each item must be defined by the
+    grid element and the element id at which it resides, e.g.:
+
+        grid_element = 'node'
+        element_id = 9.
+
+    When items are defined, each item is given a unique id and the underlying
+    Dataset uses a dimension "item_id".
+
```
Comment (kbarnhart) [inline landlab/data_record/data_record.py]: I've added some text (here and elsewhere) trying to to provide some more background on DataRecord. Let me know what you think.

File: landlab/data_record/data_record.py [hunk 2]
```diff
@@ -14,40 +15,70 @@ class DataRecord(object):
...
+    For example, if the grid has 100 links, and no dummy link values are
+    indicated, then, no item can live at link 100 or link -3 because only links
+    0 to 99 exist in this example.
+
+    Anything that the DataRecord keeps track of is considered a "record",
+    whether it uses one or both of the two standard dimensions.
+
+    DataRecord provides two method to assist with adding new records. The
+    method ``add_item`` should be used when no new variables are being added.
+    The method ``add_record`` should be used when new variables are being added.
```
Comment (kbarnhart) [inline landlab/data_record/data_record.py]: I **think** I've appropriately represented the difference between add_record and add_item. Let me know if you think it is incorrect.

File: landlab/data_record/data_record.py [hunk 3]
```diff
@@ -347,7 +383,11 @@ def _check_element_id_values(self, grid_element, element_id):
             )
 
     def add_record(self, time=None, item_id=None, new_item_loc=None, new_record=None):
-        """Add a time-related record to an existing DataRecord.
+        """Add a new record to an existing DataRecord.
+
+        Unlike add_item, this method can support adding records that include
+        new variables to the DataRecord. It can also support adding records
+        that do not include time.
```
Comment (kbarnhart) [inline landlab/data_record/data_record.py]: Similar to in the __init__ I've added text here trying to distinguish between add_record and add_item. Please comment/improve/etc.

**Comment by kbarnhart** [comment]
This is now ready for review. I've also tagged some parts of the PR in which I've attempted to improve the docstrings.

**Comment by kbarnhart** [comment]
@margauxmouchene could you take a look at this?

**Comment by kbarnhart** [comment]
@nathanlyons might you take a look at this?

**Comment by kbarnhart** [comment]
@landlab/core might a get someone to take a look at this? I'd like to merge it (or a revised version of it) for continued work with @pfeiffea

**Comment by nathanlyons** [comment]
@kbarnhart I plan to look Friday morning. I hope that's OK.

File: landlab/data_record/data_record.py [hunk 4]
```diff
@@ -14,40 +15,70 @@ class DataRecord(object):
     Dataset by providing additional attributes and functions, including the
```
[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/data_record/data_record.py]: Done

File: landlab/data_record/data_record.py [hunk 5]
```diff
@@ -14,40 +15,70 @@ class DataRecord(object):
...
     ability to aggregate values on Landlab grid elements.
 
-    Data variables can vary along one or both of the following dimensions:
+    DataRecord uses the concept of an "item", a physical thing that is located
+    on the grid and has some properties, stored as variables. DataRecord was
+    built to keep track of variables associated with a Landlab grid across
+    items, through time, or both.
```
[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/data_record/data_record.py]: Good point. Change made.

File: landlab/data_record/data_record.py [hunk 6]
```diff
@@ -14,40 +15,70 @@ class DataRecord(object):
...
             by an individual id) that reside on the grid.
 
+    If an item or set of items is defined, each item must be defined by the
+    grid element and the element id at which it resides, e.g.:
+
+        grid_element = 'node'
+        element_id = 9.
+
+    When items are defined, each item is given a unique id and the underlying
+    Dataset uses a dimension "item_id".
```
[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/data_record/data_record.py]: Good idea. Text added.

File: landlab/data_record/data_record.py [hunk 7]
```diff
@@ -2,6 +2,7 @@
 # -*- coding: utf-8 -*-
 
 import numpy as np
+import xarray as xr
```
[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/data_record/data_record.py]: Done.

File: landlab/data_record/data_record.py [hunk 8]
```diff
@@ -14,40 +15,70 @@ class DataRecord(object):
...
-    grid_element that exist on the grid. For example, if the grid has 100
-    links, no item can live at link 100 or link -3 because only links 0 to 99
-    exist in this example.
+    grid_element that exist on the grid or be one of the dummy element values.
+    For example, if the grid has 100 links, and no dummy link values are
+    indicated, then, no item can live at link 100 or link -3 because only links
+    0 to 99 exist in this example.
+
+    Anything that the DataRecord keeps track of is considered a "record",
+    whether it uses one or both of the two standard dimensions.
```
[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/data_record/data_record.py]: Added.

File: landlab/data_record/data_record.py [hunk 9]
```diff
@@ -14,40 +15,70 @@ class DataRecord(object):
...
         """
         Parameters
         ----------
         grid : ModelGrid
+        dummy_elements : dict
+             Dictionary indicating valid values for dummy grid elements. For
+             example, if you need an "exit" off of a grid with  100 links, you could indicate
+                dummy_elements = {"link": [9999]}
+             to set a link id of 9999 as a dummy link. Multiple dummy elements
+             are possible.
```
[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/data_record/data_record.py]: Good idea. Text added.

File: landlab/data_record/data_record.py [hunk 10]
```diff
@@ -95,11 +126,15 @@ def __init__(self, grid, time=None, items=None, data_vars=None, attrs=None):
...
-        variable names and values given by DataArray objects for each variable
-        name.
+        DataRecord builds off of xarray Dataset, a multi-dimensional, in
+        memory, array  database. Dataset implements the mapping interface with
+        keys given by variable names and values given by DataArray objects for
+        each variable name.
+
         A DataRecord can have dimensions 'time' and/or 'item_id'.
+
+        The xarray Dataset is stored in the attribute self.dataset.
```
[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/data_record/data_record.py]: Done.

File: landlab/data_record/data_record.py [hunk 11]
```diff
@@ -151,12 +188,22 @@ def __init__(self, grid, time=None, items=None, data_vars=None, attrs=None):
...
         """
+
         # save a reference to the grid
         self._grid = grid
 
         # depending on the grid type, permitted locations for items vary
         self._permitted_locations = self._grid.groups
 
+        # save dummy elements reference
+        # check dummies and refomat into {"node": [0, 1, 2]}
```
[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/data_record/data_record.py]: Now Fixed.

File: landlab/data_record/data_record.py [hunk 12]
```diff
@@ -347,7 +383,11 @@ def _check_element_id_values(self, grid_element, element_id):
             )
 
     def add_record(self, time=None, item_id=None, new_item_loc=None, new_record=None):
-        """Add a time-related record to an existing DataRecord.
+        """Add a new record to an existing DataRecord.
```
[the remark this answers is not in the document]
Comment (kbarnhart) [reply landlab/data_record/data_record.py]: Good idea. Done.

File: landlab/data_record/data_record.py (hunk 1 above)
Comment (kbarnhart) [reply landlab/data_record/data_record.py]: @margauxmouchene thanks for your additional comment on this point. I've added half a sentence based on what you wrote.

File: landlab/data_record/data_record.py (hunk 2 above)
Comment (kbarnhart) [reply landlab/data_record/data_record.py]: @margauxmouchene commented on this.

**Pull request merged**